modules/siddhi-extensions/markov-models/src/main/java/org/wso2/siddhi/extension/markov/MarkovChainTransitionProbabilitiesCalculator.java

```diff
@@ -0,0 +1,237 @@
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.siddhi.extension.markov;
+
+import org.apache.log4j.Logger;
+import org.wso2.siddhi.core.exception.OperationNotSupportedException;
+
+import java.io.*;
+import java.util.*;
+
+public class MarkovChainTransitionProbabilitiesCalculator implements Serializable {
+
+    private static final long serialVersionUID = -7632888859752532839L;
+    private static final Logger log = Logger.getLogger(MarkovChainTransitionProbabilitiesCalculator.class);
+    private long durationToKeep;
+    private long notificationsHoldLimit;
+    private double alertThresholdProbability;
+    private int eventCount;
+    private Map<String, String> idToLastStates = new HashMap<String, String>();
+    private Map<String, Long> idToLastStatesExpiryTime = new LinkedHashMap<String, Long>();
+    private MarkovMatrix markovMatrix = new MarkovMatrix();
+
+    /**
+     * Constructor for real time training method
+     *
+     * @param durationToKeep
+     * @param alertThresholdProbability
+     * @param notificationsHoldLimit
+     */
+    public MarkovChainTransitionProbabilitiesCalculator(long durationToKeep, double alertThresholdProbability,
+            long notificationsHoldLimit) {
+        this.durationToKeep = durationToKeep;
+        this.notificationsHoldLimit = notificationsHoldLimit;
+        this.alertThresholdProbability = alertThresholdProbability;
+    }
+
+    /**
+     * Constructor for predefined markov matrix method
+     *
+     * @param durationToKeep
+     * @param alertThresholdProbability
+     * @param markovMatrixStorageLocation
+     */
+    public MarkovChainTransitionProbabilitiesCalculator(long durationToKeep, double alertThresholdProbability,
+            String markovMatrixStorageLocation) {
+        this.durationToKeep = durationToKeep;
+        this.alertThresholdProbability = alertThresholdProbability;
+        populateMarkovMatrix(markovMatrixStorageLocation);
+    }
+
+    /**
+     * Method to process new events
+     *
+     * @param userId
+     * @param state
+     * @param trainingOption
+     * @return Array containing output results
+     */
+    protected Object[] processData(String userId, String state, boolean trainingOption) {
+
+        eventCount++;
+        String id = userId;
+        String currentState = state;
+        boolean currentTrainingOption = trainingOption;
```

Comment:
Thanks for the suggestion. Fixed with https://github.com/wso2/siddhi/pull/189/commits/690b60924a459f505cce35f57d2ce6b05e4980c2